extensions/replication/queueProcessor/QueueProcessor.js

```diff
@@ -63,24 +68,92 @@ class QueueProcessor {
         this.sourceHTTPAgent = new http.Agent({ keepAlive: true });
         this.destHTTPAgent = new http.Agent({ keepAlive: true });
 
+        this.vaultClientCache = new VaultClientCache();
+
         // FIXME support multiple destination sites
         if (destConfig.bootstrapList.length > 0) {
             this.destHosts =
                 new RoundRobin(destConfig.bootstrapList[0].servers,
                                { defaultPort: 80 });
+            if (destConfig.bootstrapList[0].echo) {
+                if (process.env.BACKBEAT_ECHO_TEST_MODE === '1') {
+                    this.logger.info('starting in echo mode',
+                                     { testMode: true });
+                } else {
+                    this.logger.info('starting in echo mode');
+                }
+                this.accountCredsCache = {};
+            }
         } else {
             this.destHosts = null;
         }
 
+        let sourceAdminVaultConfigured = false;
+        let destAdminVaultConfigured = false;
+
         if (sourceConfig.auth.type === 'role') {
-            const { host, port } = sourceConfig.auth.vault;
-            this.sourceVault = new VaultClient(host, port);
+            const { host, port, adminPort, adminCredentialsFile }
+                      = sourceConfig.auth.vault;
+            this.vaultClientCache
+                .setHost('source:s3', host)
+                .setPort('source:s3', port);
+            if (adminCredentialsFile) {
+                this.vaultClientCache
+                    .setHost('source:admin', host)
+                    .setPort('source:admin', adminPort)
```

Comment:
Yes, but rather as an error condition: the configuration would be invalid if `adminCredentialsFile` is provided but `adminPort` is not. The credentials file is necessary for echo mode, which also requires to provide the port for the admin interface. But I agree it would be good to sanity check it and abort the process with a proper error if not specified, will do something about it.

Reply:
I added this sanity check and others in the config.joi.js file, so now `adminPort` should be guaranteed to be set here.